hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkWriteHelper.java

```diff
@@ -59,10 +59,9 @@ public static SparkWriteHelper newInstance() {
     }).reduceByKey((rec1, rec2) -> {
       @SuppressWarnings("unchecked")
       T reducedData = (T) rec1.getData().preCombine(rec2.getData());
-      // we cannot allow the user to change the key or partitionPath, since that will affect
-      // everything
-      // so pick it from one of the records.
-      return new HoodieRecord<T>(rec1.getKey(), reducedData);
+      HoodieKey reducedKey = rec1.getData().equals(reducedData) ? rec1.getKey() : rec2.getKey();
+
+      return new HoodieRecord<T>(reducedKey, reducedData);
```

Comment:
Hi @rmpifer thanks for your contribution 
 It seems ` rec1.getKey()` equals  `rec2.getKey()` ? each of them is ok.

Reply:
@wangxianghu Not necessarily. If global indexed it is reduced on record key not hoodie key. In the case I mentioned `rec1` and `rec2` would have different `partitionPath` values which is why arbitrarily choosing this causes an issue.

Reply:
> @wangxianghu Not necessarily. If global indexed it is reduced on record key not hoodie key. In the case I mentioned `rec1` and `rec2` would have different `partitionPath` values which is why arbitrarily choosing this causes an issue.

I get your point, you're right. when using a global index,  we might get a wrong `HoodieKey` for the `HoodieRecord`. how about adding some comments to note this change?

BTW, please file a Jira ticket to track this change, and rename this pr to 'Fix xxx Exception when .....' ?

Reply:
> @wangxianghu Not necessarily. If global indexed it is reduced on record key not hoodie key. In the case I mentioned `rec1` and `rec2` would have different `partitionPath` values which is why arbitrarily choosing this causes an issue.

I get your point, you're right. when using a global index,  we might get a wrong `HoodieKey` for the `HoodieRecord`. how about adding some comments to note this change?
   
   BTW, please file a Jira ticket to track this change, and rename this pr to 'Fix xxx Exception when .....' ?